File: public/Own_Equipment_Tracker/js/monthly_list_summary.js
const monthNames = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

let selectedYear = null; // 🟢 Initially null (No year selected)
let selectedMonth = null; // Initially null
let currentExportType = "";

document.addEventListener("DOMContentLoaded", () => {
  if (!localStorage.getItem("eq_user")) {
    window.location.replace("index.html");
  }
  initFilters();
});

function initFilters() {
  const currentYear = new Date().getFullYear();
  const yearContainer = document.getElementById("yearButtonsContainer");
  yearContainer.innerHTML = "";

  const startYear = Math.max(currentYear, 2026);
  for (let y = startYear; y >= 2026; y--) {
    const btn = document.createElement("div");
    btn.className = `filter-btn`;
    btn.innerText = y;
    btn.onclick = () => selectYear(y);
    yearContainer.appendChild(btn);
  }
  renderMonthButtons(); // Will hide month section initially
}

// 🟢 NEW: Year Selection Logic
function selectYear(y) {
  if (selectedYear === y) {
    selectedYear = null; // Toggle Off if same year clicked
    selectedMonth = null;
  } else {
    selectedYear = y; // Set new year
    selectedMonth = null; // Reset month when year changes
  }

  // Update Year Button UI
  document
    .querySelectorAll("#yearButtonsContainer .filter-btn")
    .forEach((b) => {
      b.classList.remove("active");
      if (parseInt(b.innerText) === selectedYear) b.classList.add("active");
    });

  renderMonthButtons();
  updateTableVisibility();
}

function renderMonthButtons() {
  const monthSection = document.getElementById("monthSection");
  const container = document.getElementById("monthButtonsContainer");

  if (selectedYear === null) {
    monthSection.style.display = "none";
    container.innerHTML = "";
    return;
  }

  monthSection.style.display = "block";
  container.innerHTML = "";

  const allBtn = document.createElement("div");
  allBtn.className = `filter-btn ${selectedMonth === "ALL" ? "active" : ""}`;
  allBtn.innerText = "All Months";
  allBtn.onclick = () => selectMonth("ALL");
  container.appendChild(allBtn);

  monthNames.forEach((m, idx) => {
    const btn = document.createElement("div");
    const monthNum = idx + 1;
    btn.className = `filter-btn ${selectedMonth === monthNum ? "active" : ""}`;
    btn.innerText = m;
    btn.onclick = () => selectMonth(monthNum);
    container.appendChild(btn);
  });
}

// 🟢 NEW: Month Selection Logic
function selectMonth(m) {
  if (selectedMonth === m) {
    selectedMonth = null; // Toggle Off if same month clicked
  } else {
    selectedMonth = m; // Toggle On
  }
  renderMonthButtons();
  updateTableVisibility();
}

function updateTableVisibility() {
  const container = document.getElementById("dataContainer");
  if (selectedYear === null) {
    container.innerHTML =
      "<p style='text-align:center; font-size: 16px; color: #7f8c8d; margin-top: 50px;'>👆 Please select a Year first to proceed.</p>";
    return;
  }
  if (selectedMonth === null) {
    container.innerHTML =
      "<p style='text-align:center; font-size: 16px; color: #7f8c8d; margin-top: 50px;'>👆 Please select a month to view the data.</p>";
    return;
  }
  loadMonthlyData();
}

async function loadMonthlyData() {
  const container = document.getElementById("dataContainer");
  container.innerHTML = "<p style='text-align:center;'>Loading data...</p>";

  try {
    const res = await fetch(`/api/monthly-summary/data?year=${selectedYear}`);
    const data = await res.json();
    renderTables(data.equipments, data.logs, selectedYear, selectedMonth);
  } catch (err) {
    console.error(err);
    showToast("Error loading data", "error");
  }
}

function renderTables(equipments, logs, year, currentSelectedMonth) {
  const container = document.getElementById("dataContainer");
  container.innerHTML = "";

  let monthsToRender =
    currentSelectedMonth === "ALL"
      ? [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
      : [parseInt(currentSelectedMonth)];

  monthsToRender.forEach((m) => {
    const monthLogs = logs.filter((l) => l.month === m);
    if (monthLogs.length === 0 && currentSelectedMonth === "ALL") return;

    const wrapperId = `month-wrapper-${m}`;
    let html = `
            <div class="month-container" id="${wrapperId}">
                <div class="month-header">
                    <div style="width:100px;"></div>
                    <h3 class="month-title">Monthly Income & Expenditure Statement :: ${monthNames[m - 1]} ${year}</h3>
                    <div class="export-icons">
                        <button class="export-btn" title="Export PDF" onclick="exportPDF(${m}, '${monthNames[m - 1]}', ${year})">📄</button>
                        <button class="export-btn" title="Copy as Image" onclick="copyImage('${wrapperId}')">📋</button>
                        <button class="export-btn" title="Export Excel" onclick="exportSingleExcel(${m}, ${year})">📊</button>
                    </div>
                </div>
                
                <table id="table-${m}">
                    <thead>
                        <tr>
                            <th colspan="2" class="th-main">EQ</th>
                            <th colspan="7" class="th-main">Expenditures</th>
                            <th colspan="2" class="th-main">Income</th>
                            <th colspan="3" class="th-main">Commissions</th>
                            <th class="th-main"></th>
                        </tr>
                        <tr>
                            <th>SN</th>
                            <th class="eq-width">Plate no</th>
                            <th>EQ Maintenance Cost</th>
                            <th class="eq-width">Net Salary</th>
                            <th>Santook Rent</th>
                            <th class="eq-width">Debt</th>
                            <th class="eq-width">PWAS</th>
                            <th class="eq-width">Other</th>
                            <th class="bg-op-exp">Total Operational Expenses</th>
                            <th class="bg-op-rev">Operational revenue</th>
                            <th class="bg-net-op">Net OP Revenue</th>
                            <th class="eq-width">Kafil</th>
                            <th class="eq-width">Owner</th>
                            <th class="eq-width">Investor</th>
                            <!-- 🟢 ADDED .pl-width here -->
                            <th class="bg-pl pl-width">Profit or Loss</th>
                        </tr>
                    </thead>
                    <tbody>
        `;

    let totals = {
      maint: 0,
      netSal: 0,
      rent: 0,
      debt: 0,
      pwas: 0,
      other: 0,
      totOpExp: 0,
      opRev: 0,
      netOpRev: 0,
      kafil: 0,
      owner: 0,
      inv: 0,
      pl: 0,
    };

    equipments.forEach((eq, idx) => {
      const log = monthLogs.find((x) => x.equipment_id === eq.id) || {};
      const maint = Number(log.maintenance_cost || 0),
        basic = Number(log.basic_salary || 0),
        ot = Number(log.overtime || 0),
        penalty = Number(log.penalty || 0);
      const netSal = basic + ot - penalty;
      const rent = Number(log.santook_rent || 0),
        debt = Number(log.debit || 0),
        pwas = Number(log.pwas || 0),
        other = Number(log.other_expense || 0);
      const opRev = Number(log.op_revenue || 0);
      const totOpExp = maint + netSal + rent + debt + pwas + other;
      const netOpRev = opRev - totOpExp;
      const kafil = opRev * (Number(log.kafil_comm || 0) / 100),
        owner = Number(log.owner_comm || 0),
        inv = Number(log.investor_comm || 0);
      const totComm = kafil + owner + inv;
      const pl = netOpRev - totComm;

      totals.maint += maint;
      totals.netSal += netSal;
      totals.rent += rent;
      totals.debt += debt;
      totals.pwas += pwas;
      totals.other += other;
      totals.totOpExp += totOpExp;
      totals.opRev += opRev;
      totals.netOpRev += netOpRev;
      totals.kafil += kafil;
      totals.owner += owner;
      totals.inv += inv;
      totals.pl += pl;

      html += `
                <tr>
                    <td>${idx + 1}</td>
                    <td><b>${eq.plate_no}</b></td>
                    <td>${fmt(maint)}</td>
                    <td>${fmt(netSal)}</td>
                    <td>${fmt(rent)}</td>
                    <td>${fmt(debt)}</td>
                    <td>${fmt(pwas)}</td>
                    <td>${fmt(other)}</td>
                    <td class="bg-op-exp">${fmt(totOpExp)}</td>
                    <td class="bg-op-rev">${fmt(opRev)}</td>
                    <td class="bg-net-op">${fmt(netOpRev)}</td>
                    <td>${fmt(kafil)}</td>
                    <td>${fmt(owner)}</td>
                    <td>${fmt(inv)}</td>
                    <td class="bg-pl pl-width" style="color: ${pl < 0 ? "red" : "black"};">${fmt(pl)}</td>
                </tr>
            `;
    });

    html += `
                    <tr style="font-weight: bold; background: #fafafa;">
                        <td colspan="2">TOTAL</td>
                        <td>${fmt(totals.maint)}</td>
                        <td>${fmt(totals.netSal)}</td>
                        <td>${fmt(totals.rent)}</td>
                        <td>${fmt(totals.debt)}</td>
                        <td>${fmt(totals.pwas)}</td>
                        <td>${fmt(totals.other)}</td>
                        <td class="bg-op-exp">${fmt(totals.totOpExp)}</td>
                        <td class="bg-op-rev">${fmt(totals.opRev)}</td>
                        <td class="bg-net-op">${fmt(totals.netOpRev)}</td>
                        <td>${fmt(totals.kafil)}</td>
                        <td>${fmt(totals.owner)}</td>
                        <td>${fmt(totals.inv)}</td>
                        <td class="bg-pl pl-width" style="color: ${totals.pl < 0 ? "red" : "black"};">${fmt(totals.pl)}</td>
                    </tr>
                </tbody>
            </table>
        </div>`;
    container.innerHTML += html;
  });

  if (container.innerHTML === "") {
    container.innerHTML =
      "<p style='text-align:center;'>No operational data found for the selected period.</p>";
  }
}

function fmt(val) {
  return Number(val) === 0 ? "" : Number(val).toFixed(2);
}

// 🟢 EXPORT MODAL LOGIC
function openExportModal(type) {
  currentExportType = type;

  let monthOptions = "";
  monthNames.forEach((m, idx) => {
    monthOptions += `<option value="${idx + 1}">${m}</option>`;
  });

  let yearOptions = "";
  const currentYear = new Date().getFullYear();
  const startYear = Math.max(currentYear, 2026);
  for (let y = 2026; y <= startYear; y++) {
    yearOptions += `<option value="${y}">${y}</option>`;
  }

  document.getElementById("fromMonth").innerHTML = monthOptions;
  document.getElementById("toMonth").innerHTML = monthOptions;
  document.getElementById("fromYear").innerHTML = yearOptions;
  document.getElementById("toYear").innerHTML = yearOptions;

  // Use current year if selectedYear is null
  const defaultYear = selectedYear || new Date().getFullYear();
  document.getElementById("fromYear").value = defaultYear;
  document.getElementById("toYear").value = defaultYear;

  if (selectedMonth !== "ALL" && selectedMonth !== null) {
    document.getElementById("fromMonth").value = selectedMonth;
    document.getElementById("toMonth").value = selectedMonth;
  } else {
    document.getElementById("fromMonth").value = 1;
    document.getElementById("toMonth").value = 12;
  }

  document.getElementById("exportModal").style.display = "flex";
}

function closeExportModal() {
  document.getElementById("exportModal").style.display = "none";
}

async function executeBatchExport() {
  const fromM = parseInt(document.getElementById("fromMonth").value);
  const fromY = parseInt(document.getElementById("fromYear").value);
  const toM = parseInt(document.getElementById("toMonth").value);
  const toY = parseInt(document.getElementById("toYear").value);

  if (fromY > toY || (fromY === toY && fromM > toM)) {
    showToast(
      "Invalid Date Range! 'From' date must be before 'To' date.",
      "error",
    );
    return;
  }

  closeExportModal();

  if (currentExportType === "excel") {
    window.location.href = `/py/monthly-summary/export-excel?from_y=${fromY}&from_m=${fromM}&to_y=${toY}&to_m=${toM}`;
    showToast("Generating Excel...", "success");
  } else {
    showToast("Generating Batch PDF, please wait...", "success");
    const prevYear = selectedYear;
    const prevMonth = selectedMonth;

    try {
      let periods = [];
      for (let y = fromY; y <= toY; y++) {
        let start = y === fromY ? fromM : 1;
        let end = y === toY ? toM : 12;
        for (let m = start; m <= end; m++) {
          periods.push({ year: y, month: m });
        }
      }

      const yearsToFetch = [...new Set(periods.map((p) => p.year))];
      let allData = {};
      for (let y of yearsToFetch) {
        const res = await fetch(`/api/monthly-summary/data?year=${y}`);
        allData[y] = await res.json();
      }

      for (let p of periods) {
        const data = allData[p.year];
        renderTables(data.equipments, data.logs, p.year, p.month);
        await new Promise((r) => setTimeout(r, 400));

        const element = document.getElementById(`month-wrapper-${p.month}`);
        if (element) {
          const exportIcons = element.querySelector(".export-icons");
          if (exportIcons) exportIcons.style.display = "none";

          const originalOverflow = element.style.overflowX;
          element.style.overflowX = "visible";

          const HTML_Width = element.scrollWidth;
          const HTML_Height = element.scrollHeight;

          const opt = {
            margin: 10,
            filename: `Income_Expenditure_${monthNames[p.month - 1]}_${p.year}.pdf`,
            image: { type: "jpeg", quality: 1.0 },
            html2canvas: { scale: 2, useCORS: true, windowWidth: HTML_Width },
            pagebreak: { mode: "avoid-all" },
            jsPDF: {
              unit: "px",
              format: [HTML_Width + 20, HTML_Height + 50],
              orientation: "landscape",
            },
          };

          await html2pdf().set(opt).from(element).save();
        }
      }

      selectedYear = prevYear;
      selectedMonth = prevMonth;
      renderMonthButtons();
      updateTableVisibility();
      showToast("Batch PDF Export Completed!", "success");
    } catch (e) {
      showToast("Error generating PDFs", "error");
      console.error(e);
    }
  }
}

function exportSingleExcel(month, year) {
  window.location.href = `/py/monthly-summary/export-excel?from_y=${year}&from_m=${month}&to_y=${year}&to_m=${month}`;
  showToast("Generating Excel...", "success");
}

function exportPDF(month, monthName, year) {
  const element = document.getElementById(`month-wrapper-${month}`);
  const exportIcons = element.querySelector(".export-icons");
  if (exportIcons) exportIcons.style.display = "none";

  const originalOverflow = element.style.overflowX;
  element.style.overflowX = "visible";

  const HTML_Width = element.scrollWidth;
  const HTML_Height = element.scrollHeight;

  const opt = {
    margin: 10,
    filename: `Income_Expenditure_${monthName}_${year}.pdf`,
    image: { type: "jpeg", quality: 1.0 },
    html2canvas: { scale: 2, useCORS: true, windowWidth: HTML_Width },
    pagebreak: { mode: "avoid-all" },
    jsPDF: {
      unit: "px",
      format: [HTML_Width + 20, HTML_Height + 50],
      orientation: "landscape",
    },
  };

  html2pdf()
    .set(opt)
    .from(element)
    .save()
    .then(() => {
      if (exportIcons) exportIcons.style.display = "flex";
      element.style.overflowX = originalOverflow;
      showToast("PDF Exported Successfully!", "success");
    });
}

function copyImage(wrapperId) {
  const element = document.getElementById(wrapperId);
  const exportIcons = element.querySelector(".export-icons");
  if (exportIcons) exportIcons.style.display = "none";

  const originalOverflow = element.style.overflowX;
  element.style.overflowX = "visible";

  html2canvas(element, { scale: 3, backgroundColor: "#ffffff" }).then(
    (canvas) => {
      if (exportIcons) exportIcons.style.display = "flex";
      element.style.overflowX = originalOverflow;

      canvas.toBlob((blob) => {
        try {
          const item = new ClipboardItem({ "image/png": blob });
          navigator.clipboard
            .write([item])
            .then(() => {
              showToast("Image copied to clipboard!", "success");
            })
            .catch(() => downloadImage(canvas, "Statement.png"));
        } catch (e) {
          downloadImage(canvas, "Statement.png");
        }
      }, "image/png");
    },
  );
}

function downloadImage(canvas, filename) {
  const a = document.createElement("a");
  a.href = canvas.toDataURL("image/png");
  a.download = filename;
  a.click();
  showToast("Image Downloaded!", "success");
}

function showToast(message, type = "success") {
  const container = document.getElementById("toast-container");
  const toast = document.createElement("div");
  toast.className = `toast ${type}`;
  toast.innerHTML = `<span>${type === "success" ? "✅ " : "❌ "} ${message}</span>`;
  container.appendChild(toast);
  setTimeout(() => toast.classList.add("show"), 10);
  setTimeout(() => {
    toast.classList.remove("show");
    setTimeout(() => toast.remove(), 300);
  }, 3000);
}

File: routers/monthly_list_summary.py
import os
import io
import psycopg2
from fastapi import APIRouter, HTTPException
from fastapi.responses import Response
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter

router = APIRouter(prefix="/py/monthly-summary", tags=["Monthly Summary Export"])

@router.get("/export-excel")
async def export_monthly_summary_excel(from_y: int, from_m: int, to_y: int, to_m: int):
    try:
        conn = psycopg2.connect(
            dbname=os.getenv("DB_NAME"), user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASS"), host=os.getenv("DB_HOST", "127.0.0.1"), port=os.getenv("DB_PORT", "5432")
        )
        cur = conn.cursor()
        
        cur.execute("SELECT id, plate_no FROM equipments ORDER BY id ASC")
        equipments = cur.fetchall()
        
        # 🟢 SQL query to fetch data between From and To Date Range
        query = """
            SELECT * FROM equipment_monthly_logs 
            WHERE (year > %s OR (year = %s AND month >= %s))
              AND (year < %s OR (year = %s AND month <= %s))
            ORDER BY year ASC, month ASC
        """
        cur.execute(query, (from_y, from_y, from_m, to_y, to_y, to_m))
            
        logs = cur.fetchall()
        columns = [desc[0] for desc in cur.description]
        logs_dicts = [dict(zip(columns, row)) for row in logs]
        
        cur.close()
        conn.close()

        output = io.BytesIO()
        wb = openpyxl.Workbook()
        wb.remove(wb.active)

        month_names = ["", "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
        
        bold_font = Font(name="Calibri", size=11, bold=True)
        center_align = Alignment(horizontal="center", vertical="center", wrap_text=True)
        thin_border = Border(left=Side(style='thin'), right=Side(style='thin'), top=Side(style='thin'), bottom=Side(style='thin'))
        
        fill_op_exp = PatternFill(start_color="F8CBAD", end_color="F8CBAD", fill_type="solid")
        fill_op_rev = PatternFill(start_color="C6E0B4", end_color="C6E0B4", fill_type="solid")
        fill_net_op = PatternFill(start_color="BDD7EE", end_color="BDD7EE", fill_type="solid")
        fill_pl = PatternFill(start_color="E8D0E8", end_color="E8D0E8", fill_type="solid")

        # 🟢 Generate Sheets for the Date Range
        for y in range(from_y, to_y + 1):
            start_month = from_m if y == from_y else 1
            end_month = to_m if y == to_y else 12
            
            for m in range(start_month, end_month + 1):
                month_logs = [l for l in logs_dicts if l['year'] == y and l['month'] == m]
                # Skip empty months in batch exports, except when explicitly exporting a single month
                if not month_logs and not (from_y == to_y and from_m == to_m):
                    continue
                
                ws = wb.create_sheet(title=f"{month_names[m]} {y}")
                
                # Title
                ws.merge_cells("A1:O1")
                title_cell = ws.cell(row=1, column=1, value=f"Monthly Income & Expenditure Statement :: {month_names[m]} {y}")
                title_cell.font = Font(size=14, bold=True)
                title_cell.alignment = center_align
                ws.row_dimensions[1].height = 25
                
                # Header Row 1
                ws.merge_cells("A2:B2"); ws.cell(row=2, column=1, value="EQ").alignment = center_align
                ws.merge_cells("C2:I2"); ws.cell(row=2, column=3, value="Expenditures").alignment = center_align
                ws.merge_cells("J2:K2"); ws.cell(row=2, column=10, value="Income").alignment = center_align
                ws.merge_cells("L2:N2"); ws.cell(row=2, column=12, value="Commissions").alignment = center_align
                ws.cell(row=2, column=15, value="").alignment = center_align
                
                for col in range(1, 16):
                    ws.cell(row=2, column=col).font = bold_font
                    ws.cell(row=2, column=col).border = thin_border
                
                # Header Row 2
                headers = ["SN", "Plate no", "EQ Maintenance Cost", "Net Salary", "Santook Rent", "Debt", "PWAS", "Other", 
                           "Total Operational Expenses", "Operational revenue", "Net OP Revenue", "Kafil", "Owner", "Investor", "Profit or Loss"]
                
                ws.row_dimensions[3].height = 40
                for col_idx, text in enumerate(headers, 1):
                    cell = ws.cell(row=3, column=col_idx, value=text)
                    cell.font = bold_font
                    cell.border = thin_border
                    cell.alignment = center_align
                    
                    if col_idx == 9: cell.fill = fill_op_exp
                    elif col_idx == 10: cell.fill = fill_op_rev
                    elif col_idx == 11: cell.fill = fill_net_op
                    elif col_idx == 15: cell.fill = fill_pl
                    
                # Data Rows
                r_idx = 4
                for i, eq in enumerate(equipments, 1):
                    log = next((item for item in month_logs if item["equipment_id"] == eq[0]), {})
                    
                    maint = float(log.get("maintenance_cost", 0) or 0)
                    basic = float(log.get("basic_salary", 0) or 0)
                    ot = float(log.get("overtime", 0) or 0)
                    penalty = float(log.get("penalty", 0) or 0)
                    net_sal = (basic + ot) - penalty
                    rent = float(log.get("santook_rent", 0) or 0)
                    debt = float(log.get("debit", 0) or 0)
                    pwas = float(log.get("pwas", 0) or 0)
                    other = float(log.get("other_expense", 0) or 0)
                    
                    op_rev = float(log.get("op_revenue", 0) or 0)
                    
                    tot_op_exp = maint + net_sal + rent + debt + pwas + other
                    net_op_rev = op_rev - tot_op_exp
                    
                    kafil = op_rev * (float(log.get("kafil_comm", 0) or 0) / 100)
                    owner = float(log.get("owner_comm", 0) or 0)
                    inv = float(log.get("investor_comm", 0) or 0)
                    
                    tot_comm = kafil + owner + inv
                    pl = net_op_rev - tot_comm
                    
                    v = lambda x: "" if x == 0 else round(x, 2)
                    
                    row_data = [i, eq[1], v(maint), v(net_sal), v(rent), v(debt), v(pwas), v(other), 
                                v(tot_op_exp), v(op_rev), v(net_op_rev), v(kafil), v(owner), v(inv), v(pl)]
                    
                    ws.row_dimensions[r_idx].height = 25
                    
                    for c_idx, val in enumerate(row_data, 1):
                        cell = ws.cell(row=r_idx, column=c_idx, value=val)
                        cell.border = thin_border
                        cell.alignment = center_align
                        if c_idx == 9: cell.fill = fill_op_exp
                        elif c_idx == 10: cell.fill = fill_op_rev
                        elif c_idx == 11: cell.fill = fill_net_op
                        elif c_idx == 15: 
                            cell.fill = fill_pl
                            if pl < 0: cell.font = Font(color="FF0000")
                    r_idx += 1
                    
                # Formatting Columns width
                equal_width_cols = [2, 4, 6, 7, 8, 12, 13, 14] 
                for col in range(1, 16):
                    if col in equal_width_cols:
                        ws.column_dimensions[get_column_letter(col)].width = 15 
                    elif col == 15: # 🟢 Profit or Loss Column Width
                        ws.column_dimensions[get_column_letter(col)].width = 18 
                    else:
                        ws.column_dimensions[get_column_letter(col)].width = 14

        if len(wb.sheetnames) == 0:
            wb.create_sheet("No Data")
            
        wb.save(output)
        output.seek(0)
        
        if from_y == to_y and from_m == to_m:
            filename = f"Monthly_Statement_{month_names[from_m]}_{from_y}.xlsx"
        else:
            filename = f"Statement_Batch_{month_names[from_m]}_{from_y}_To_{month_names[to_m]}_{to_y}.xlsx"
            
        return Response(
            content=output.getvalue(),
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))